### PR TITLE
🚨🚨 Fix beam score calculation issue for decoder-only models

### LINKED ISSUE #26624: Beam search calculates mean logprobs wrong?
### System Info

- `transformers` version: 4.33.3
- Platform: Linux-5.15.0-76-generic-x86_64-with-glibc2.31
- Python version: 3.10.13
- Huggingface_hub version: 0.17.3
- Safetensors version: 0.3.3
- Accelerate version: 0.23.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@gante (recommended for generate-related issues)
@patrickvonplaten (wrote the code according to git-blame)

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

device = "cuda"
model_name = "gpt2"
short_prompt = "Once upon a time"
long_prompt = """In a hole in the ground there lived a hobbit. Not a nasty, dirty, wet hole, filled with the ends of bworms and an oozy smell, nor yet a dry, bare, sandy hole with nothing in it to sit down on or to eat: it a was a hobbit-hole, and that means comfort.
It had a perfectly round door like a porthole, painted green, with a shiny yellow brass knob in the exact middle. The door opened on to a tube-shaped hall like a tunnel: a very comfortable tunnel without smoke, with panelled walls, and floors tiled and carpeted, provided with polished chairs, and lots and lots of pegs for hats and coats- the hobbit was fond of visitors. The tunnel wound on and on – going fairly but not quite straight into the side of the hill – The Hill, as all the people for many miles around called it – and many little round doors opened out of it, first on one side and then on another. No going upstairs for the hobbit: bedrooms, bathrooms, cellars, pantries (lots of these), wardrobes (he had whole rooms devoted to clothes), kitchens, dining-rooms, all were on the same floor, and indeed on the same passage. The best rooms were all on the lefthand side (going in), for these were the only ones to have windows, deep-set round windows looking over his garden, and meadows beyond, sloping down to the river.
This hobbit was a very well-to-do hobbit, and his name was Baggins. The Bagginses have lived in the neighbourhood of The Hill for time out of mind, and people considered them very respectable, not only because most of them were rich, but also because they never had any adventures or did anything unexpected: you could tell what a Baggins would say on any question without the bother of asking him. This is a story of how a Baggins had an adventure, and"""

beam_size = 5
max_new_tokens = 1
num_return_sequences = beam_size

with torch.device(device), torch.no_grad():
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForCausalLM.from_pretrained(model_name)

    for prompt_name, prompt in [("short_prompt", short_prompt), ("long_prompt", long_prompt)]:
        batch = tokenizer(prompt, return_tensors="pt")
        outputs = model.generate(**batch, num_beams=beam_size, num_return_sequences=num_return_sequences,
                                 max_new_tokens=max_new_tokens, output_scores=True, return_dict_in_generate=True)
        print(f"{prompt_name}: scores={outputs.sequences_scores.tolist()}")
```

Prints:
```
short_prompt: scores=[-0.17024032771587372, -0.5479289293289185, -0.6405749320983887, -0.6600505113601685, -0.7051623463630676]
long_prompt: scores=[-0.0049241515807807446, -0.006088315974920988, -0.006767737679183483, -0.006866625044494867, -0.006999899633228779]
```

### Expected behavior

When doing beam search, beam scores are normalized to represent the average token logprob. However, the current implementation divides the sum of **_generated token logprobs_** by the length of the **_entire sequence, including prompt_**. This creates inconsistencies between the scores of sequences of different lengths, and also prefers shorter generations. [Code here](https://github.com/huggingface/transformers/blob/75a33d60f25d99ff8cdd657d6ba685dc4336a0d1/src/transformers/generation/beam_search.py#L938).

My reproduction example shows that the absolute values of beam scores returned by generating a single token with a long prompt are orders of magnitude smaller than beam scores returned by a short prompt.

The main scenario where this behavior is problematic is for beams that terminate with an EOS before `max_new_tokens` is reached, since the denominator in their score calculation will be skewed. For example, if we have 2 candidates with lengths `l1` and `l2`, where all token logprobs are `s` and the prompt length is `p`, we'll have:
`score_i = l_i * s / (l_i + p) = s / (1 + p / l_i)`, showing a preference for shorter generations since `s < 0`.

### PR DESCRIPTION
**Pull request by VsonicV**
# What does this PR do?

This PR fixes issue #26624 . In the original implementation of beam search, the beam score for decoder-only models is normalized by the total length of both prompt and generated sequence. However, the length of prompt should not be included in the normalization step. This issue would cause an unexpected bias towards generating shorter sequences.

This is a simple quick fix by adding an optional parameter `decoder_prompy_len`, which stores the length of prompt in decoder, to `BeamSearchScorer.process()`, `BeamSearchScorer.finalize()` and `BeamHypotheses.add()`. Since the added new parameter is optional with a default value as 0, any existing calls of these functions without specifying `decoder_prompy_len` would still work in the same way as before, avoiding any unexpected incompatibility. The corner case in which the very first generated token happens to be eos_token (empty generation) is considered and handled.

Fixes #26624 

Note: There are three follow-up PRs that complement this fix:
1. #27808 further fixes some remaining issues in the Pytorch version.
2. #27814 fixes the Tensorflow version.
3. #27816 fixes the JAX version.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@gante

### CODE REVIEW

**Comment by VsonicV** [comment]
@gante This fix is only for `beam_search`. If you think the fix is good to go, I can also apply the same fix to `beam_sample`,  `group_beam_search` and `constrained_beam_search`.

**Comment by VsonicV** [comment]
@gante I think the current tests regarding `beam_search` are using the results generated by previous "buggy" version, so the new `beam_search` cannot pass the test `test_eos_token_id_int_and_list_beam_search`, which uses the decoder-only GPT-2. We need to update the relevant tests as well.

**Review by gante — APPROVED** [review]
LGTM, thank you for jumping in to fix it!

Regarding the GPT-2 tests: agreed, they should be updated.

**Comment by gante** [comment]
@VsonicV the "setup and quality" CI can be fixed by running `make fixup` on your local `transformers` root folder and committing the changes!

**Comment by VsonicV** [comment]
@gante Thanks for the suggestion! I have fixed the code quality issue using `make fixup`, and updated the relevant test `test_eos_token_id_int_and_list_beam_search` with new `expectation` value. Both checks pass now. However, there is still one check failure caused by `test_run_image_classification_no_trainer` and `test_run_ner_no_trainer`, which should be irrelevant to these commits regarding beam search. Do you have any clue about how to fix it?

**Comment by gante** [comment]
@VsonicV perfect! The failing CI is indeed unrelated (fix: https://github.com/huggingface/transformers/pull/27353), the tests should pass after it gets merged.

To keep the consistency of this fix throughout beam methods, I'd like to request you to:
1. Also apply this change to other beam methods in Pytorch :)
2. (optional, only if you're comfortable with it, as the fix is slightly different) Apply this change to beam methods in TF and JAX

After 1. is done, I'll tag a core maintainer to greenlight the merge!

**Comment by VsonicV** [comment]
@gante Sure! I will work on 1 and 2 in the next 2 days. Will try to do 3 after that.

**Comment by gante** [comment]
The PR causing the CI to fail was merged, and I was informed that current PRs will need to be rebased to pass CI 🤗

**Comment by VsonicV** [comment]
@gante Item 1 and 2 are done! I have applied the fix to all beam related methods: `beam_sample`, `group_beam_search` and `constrained_beam_search`. I have rebased the PR and all relevant tests have passed.

Regarding the remaining check failures, the recent merge only fixes the check failure caused by `test_run_image_classification_no_trainer`, but not for `test_run_ner_no_trainer`. According to the error message `AssertionError: 0.5109489440917969 not less than 0.5`, the checking threshold for `self.assertLess(result["train_loss"], 0.5)` in `test_run_ner_no_trainer` needs to be adjusted as well. Moreover, one new check failure is caused by `test_cached_model_has_minimum_calls_to_head` and `test_cached_tokenizer_has_minimum_calls_to_head`, which are unrelated to the commits in this PR (we only see this after the most recent rebase).

**Comment by gante** [comment]
@VsonicV yes, we are still having some CI failures (unrelated to this PR) 😭

**Comment by VsonicV** [comment]
@gante Tried rebasing once more, all the previous check failures are gone, but got one new CI failure caused by `test_assisted_decoding_sample`, which should again be unrelated to this PR.

**Comment by VsonicV** [comment]
@ArthurZucker I have rebased this PR with all your recently added test skips, the CI failures caused by `test_assisted_decoding_sample` still persist for `blenderbot`, same failures also happened for `pegasus` and `umt5` in my previous tries, would you mind adding skips of `test_assisted_decoding_sample` for `blenderbot`, `pegasus` and `umt5` as well? Thank you!

**Comment by ArthurZucker** [comment]
Yeah I'll skip this test for everyone this is getting annoying! 😅

File: tests/generation/test_framework_agnostic.py
Comment (gante) [inline tests/generation/test_framework_agnostic.py:636]:
```suggestion
        if is_pt:
            expectation = 20
        else:
            # TODO (joao): fix me
            expectation = 13
```

This test will likely fail on TF, since we haven't applied this upgrade there. Let's add a TODO for now

**Comment by gante** [comment]
Tagging @amyeroberts for a final check

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

File: src/transformers/generation/beam_search.py
Comment (VsonicV) [reply src/transformers/generation/beam_search.py:522]: good catch, will do

**Comment by VsonicV** [comment]
@gante @amyeroberts All your suggested changes have been added and committed. All the tests have passed now (finally!). Should be ready for merge.

**Comment by VsonicV** [comment]
@gante No problem! Regarding the fix of TF and JAX version, I have looked at the relevant codes briefly, and I think I can fix them. I will try to submit another PR fixing both TF and JAX later this week.

**Comment by VsonicV** [comment]
> @gante No problem! Regarding the fix of TF and JAX version, I have looked at the relevant codes briefly, and I think I can fix them. I will try to submit another PR fixing both TF and JAX later this week.

@gante Sorry about the delay in the next steps. I had a severe flu last week and just recovered. Will start working on the remaining fixes.

**Comment by VsonicV** [comment]
@gante Hi, I noticed that in the recent release notes of v4.36.0, only this PR is listed in "Beam score calculation for decoder-only models" section under "Breaking changes". Should we also add the 3 follow-up PRs ( #27808 #27814 #27816 ) under that section? It would be more clear for people to check all the changes relevant to this breaking change. Thanks.

**Comment by gante** [comment]
@VsonicV updated the release notes for future reference 👍 Thank you for your suggestion

**Pull request merged**
